### ISSUE TITLE
Use the nimble stack instead of current bluetooth stack

### ISSUE THREAD

**Issue by bluelu** [esphome/feature-requests#810]
**Describe the problem you have/What new integration you would like**

The current bluetooth implementation uses the majority of available flash and memory space.  
Someone has backported the new nimble stack to arduino (https://github.com/h2zero/NimBLE-Arduino), with only minor differences to the API.

It would be great if all the bluetooth classes could be ported to use this library instead of the current implementation.

**Please describe your use case for this integration and alternatives you've tried:**

**Additional context**

**Comment by OttoWinter** [comment]
ESPHome doesn't use the Arduino BLE stack.

We directly access the functions from esp-idf. From a quick look this "NimBLE" stack also uses esp-idf internally.

**Comment by bluelu** [comment]
@OttoWinter I might be wrong, but I thought that this component would use the nimble stack 
 (backported from idf-4.0) directly without using the old direct access functions from esp-idf 3.x based on the bluedroid stack (https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/bluetooth/esp_gap_ble.html), and thus save memory?

There is also a non Arduino component version at https://github.com/h2zero/esp-nimble-cpp

However, it's only available for esp32.

I did test it by replacing the ble_scanner platform by the sample code from the example, and I saved 70KB of memory usage (showed by the esphome debug module), while still publishing all discovered devices over a text sensor field.

**Comment by OttoWinter** [comment]
Ok I thought NimBLE was a library like Arduino's BLE library by neil kolban.

But as far as I understand now nimble is actually an esp-idf component. Is that right?

If so, that would mean we only have to replace the function calls to the bluetooth stack in `esp32_ble_scanner.cpp` from `esp_bluedroid_` to `esp_nimble_` variants (obviously the APIs won't be 100% the same).

**Comment by bluelu** [comment]
Yes, nimble is an esp-idf component and has been ported to esp32 by espressif, but only available from esp-idf 4.0+ on. The arduino component is a backport of that version to earlier esp-idfs, and https://github.com/h2zero/esp-nimble-cpp seems to provide the old bluedroid stack API calls through nimble (however with a few changes).

Yes, but esp32 ble scanner depends on esp32_ble_tracker. I think it would be best to do the replacement for all the bluetooth classes (at least for esp32).

If esphome uses idf > 4.0, the best would be to use the esp-idf component directly I assume, otherwise the arduino component (which according to https://github.com/h2zero/NimBLE-Arduino/issues/80 will be published soon in platform.io) which should also contain the base nimble APi calls which could be used instead of the arduino API.

**Comment by AieatAssam** [comment]
Backport is already available for ESP-IDF 3.2 and 3.3:
https://github.com/h2zero/esp-nimble-component

**Comment by morganrallen** [comment]
The NimBLE BLE stack is fantastic, I use it directly with the IDF in several projects, would be great to see ESPHome move over as I suspect it'll solve a lot of these WiFi disconnect issues.

**Comment by hawkeye217** [comment]
Encountering wifi disconnect issues pretty routinely (almost exactly 15 minutes apart each time) with an ESP32 controlling some LEDs and using ```esp32_ble_tracker:`` for 4 Xiaomi sensors.  It's annoying because the LEDs flash briefly every time the disconnect happens :(

**Comment by reaper7** [comment]
@hawkeye217 - I have similar, bad experiences :(
as I can see, many people have a problem with the current ble implementation 
even if their projects are not very extensive...

maybe it's worth rethinking switching to NimBLE?

**Comment by met67** [comment]
Maybe now that EspHome moved under Home Assistant umbrella they will find the reources to try this? I can't have BLE tracker and Display at the same time...

**Comment by oetelaar24** [comment]
I'd also like to vote this up.. Seems like BLE is quite hefty right now and if there is any way of improving this, it would be very nice.
Running into same memory issues on a ESP32_CAM module trying to do BLE and camera at the same time...

**Comment by AieatAssam** [comment]
I have migrated my BLE monitors to ISP-IDF and am noticing a marked improvement in flash utilisation - from 85%+ down to 60%, allowing me to enable display support. As of right now, I have TTGO T-Display module polling 3 different Xiaomi LYWSD03MMC devices (pvvx firmware reflashed) and displaying the results on screen. Even with all these components (and WiFi) enabled, I am in reasonable utilisation ranges:
![image](https://user-images.githubusercontent.com/7803161/142390109-24994f07-0548-4388-a29b-6da3752a84e7.png)

Wonder how much further this can be pushed with NimBLE. But even as it stands, ISP-IDF is a great improvement over Arduino-based implementation. 

Running stable for 3 days+ as of 18 Nov 2021.

**Comment by geiseri** [comment]
this is good news!  I will see if I can move my m5sticks this weekend.

**Comment by rradar** [comment]
@AieatAssam I wonder if you got a improved reception of ble packets? Typically the Xiaomi LYWSD03MMC publish temp+humidty every 10 seconds. So far with "default" esphome setup more than 90% of this packages got lost and esphome could only capture a fraction of it.

Does your configuration with ISP-IDF improve this situation?

**Comment by nagyrobi** [comment]
https://github.com/pvvx/ATC_MiThermometer/issues/149#issuecomment-1072981933

**Comment by mvgijssel** [comment]
I've started to experiment with the NimBLE stack in ESPHome here https://github.com/mvgijssel/setup/pull/127 by creating two external components `nimble_tracker` and `nimble_rssi`. This enables me to reliably track Apple watches which is not possible with Bluedroid at the moment. 

Also I'm kinda surprised how stable the NimBLE stack is (famous last words 😅 )!

**Comment by nagyrobi** [comment]
Is it an 1:1 replacement? Does it work with `bluetooth_proxy`?
How's your memory and heap usage compared to ESPHome's default?

**Comment by mvgijssel** [comment]
> Is it an 1:1 replacement?

I think for the most part it can be! For my use case the rpa/irk is good enough, but the other functionality should be relatively easy to implement now that there is a start.

 > Does it work with `bluetooth_proxy`? 
 Don’t know! If the Bluetooth proxy feature relies on Bluetooth serial (or classic) then no.

> Is it `esp-idf` only?
Currently it’s esp-idf only, but the libraries I’m using also support Arduino.

> How's your memory and heap usage compared to ESPHome's default?
Haven’t checked this! Will do once I continue hacking and post back here :)

**Comment by nagyrobi** [comment]
Problem with `esp-idf` is that most of the components in ESPHome are not (yet) working on that platform, especially Ethernet, which is a popular feature these days.

**Comment by benedikt-bartscher** [comment]
@nagyrobi Ethernet with esp-idf works fine now: https://github.com/esphome/feature-requests/issues/1683

**Comment by nagyrobi** [comment]
Yes, I know and it's much better than in the old days.

**Comment by andrewjswan** [comment]
I'd also like to vote this up!
Seems like BLE is quite hefty right now and if there is any way of improving this, it would be very nice.

**Comment by h2zero** [comment]
I'm surprised that this is still open. I'd be happy to help refactor the BLE implementation in esphome to use the NimBLE stack if interested, email is in my profile.

**Comment by andrewjswan** [comment]
After the update to 2025.7, all devices using the Arduino framework and Bluetooth stopped flashing due to the huge size of the firmware (105-110% of the size).
It turns out that using Bluetooth in ESPHome with the Arduino framework has become almost impossible.
Maybe there is still a way to update Bluetooth in ESPHome for the Arduino framework?

**Comment by ssieb** [comment]
Can't you use esp-idf instead?

**Comment by andrewjswan** [comment]
> Can't you use esp-idf instead?

To do this, you need to flash the ESP via USB, because simply switching the framework from Arduino to IDF either leads to a rollback of the firmware to the old one or to the loss of the ESP, it no longer boots.

**Comment by ssieb** [comment]
It works fine.  Many people have done it.  If you get the rollback, then flash it twice in a row.

**Comment by andrewjswan** [comment]
> It works fine. Many people have done it. If you get the rollback, then flash it twice in a row.

Well, how can I say, I re-flashed one ESP 5 times, there was always a rollback to the old version, i.e. to 2025.6.3, the second ESP just disappeared and never got connected again. After I flashed it via USB, everything works. But I wouldn't say the OTA option is stable and problem-free.

**Comment by andrewjswan** [comment]
Well, the Arduino framework in some cases works more stably than ESP-IDF.

**Comment by fl4p** [comment]
I started writing a proof of concept firmware, and it works very well. It just implements the basics of the esphome api and uses NimBLE.   It ended up including a BLE device repeater too, and a WiFi NAT router. 

https://github.com/fl4p/nimble-ble-proxy-esphome

I will try to fire up a PR.

**Comment by mikesloper** [comment]
I developed two firmwares one using esphome's Bluedroid implementation and one using nimble in order to compare memory usage. these were the results:


Metric | Before | After | Delta
-- | -- | -- | --
Free internal heap | ~379 KB | ~384 KB | ~+5 KB
Min free heap | ~369 KB | ~375 KB | ~+6 KB
Fragmentation | 17.9% | 18.9% | ~same
Free PSRAM | ~31.2 MB | ~31.5 MB | ~same

not sure if a 5KB saving in internal memory is worth it? it could mean the difference between having 3 BLE connexions with Blue Droid and 4 BLE connexions with nimble? It's worth noting that the Nimble implementation was not a complete port but simply a custom component using the nimble API. also mostly vibe coded.

**Comment by AzonInc** [comment]
I mean 5KB is a lot speaking about microcontrollers

**Comment by h2zero** [comment]
Been a while since this started but these days it fully depends on the chip/platform used. If using arduino code base then the current core is using NimBLE for all chips except the original esp32. There would be a drastic difference if this was a proper comparison IMO, maybe I'll give it a test.

**Comment by mikesloper** [comment]
yes this was on a ESP32P4 with C6 hosted. That's probably why it didn't make much of a difference
